6_api/README.md

```diff
@@ -1,5 +1,237 @@
+                author: 
+                {
+                    value: "<em>pg</em>",
+                    matchLevel: "full",
+                    matchedWords: 
+                    [
+                        "pg"
+                    ]
+                }
+            }
+        }, [...]
+    ],
+    page: 0,
+    nbHits: 11,
+    nbPages: 1,
+    hitsPerPage: 20,
+    processingTimeMS: 1,
+    query: "pg",
+    params: "query=pg"
+
+}
+```
+
+We need to parse the body of the HTTP response provided as Json String into our data structure in Rust. 
+
+=> This a `Deserialisation` 
+
+## Serialisation / deserialization
+
+[Serde](https://serde.rs/) is a framework for serializing and deserializing Rust data structures efficiently and generically.
+
+
+Serde provides a derive macro to generate implementations of the Serialize and Deserialize traits for data structures defined in your crate, allowing them to be represented conveniently in all of Serde's data formats.
+
+Add Serde dependency with derive feature : 
+
+```bash
+cargo add serde -F derive
+```
+
+Our data structure is composed of numbers of hits and a collection of articles
+
+We need to define our data structure representing the Json Response : 
+
+```rust
+#[derive(Debug, Deserialize)]
+pub struct Article {
+    pub title: String,
+}
+```
+
+:bulb: we don't need to map every field. Others fields are ignored.
+
+```rust
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SearchResult {
+    pub nb_hits: u64,
+    pub hits: Vec<Article>,
+}
+
+```
+
+:bulb: We use a macro `rename_all` to rename deserialized string from a `camelCase` convention to a Rust `snake_case` convention
+
+
+
+:books: Additional resources
+
+* [Serde documentation](https://doc.rust-lang.org/cargo/commands/cargo-add.html) (ex: specify version, features,etc)
+* [curl documentation](https://github.com/curl/curl)
+
+
+## Make an HTTP request
+
+The [reqwest](https://docs.rs/reqwest/latest/reqwest/) crate provides a convenient, higher-level HTTP Client. 
+
+We need to add our library to our project :
+
+```basH
+cargo add reqwest -F json blocking
+```
+
+:bulb: Reqwest supports async and blocking requests. For now, we use a blocking client.
+
+
+Example of GET HTTP request : 
+
+```rust
+// format valid URL
+let query = format!("http://hn.algolia.com/api/v1/search?query={}", keyword);
+// execute HTTP call
+let http_response: Result<Result, Error> = reqwest::blocking::get(query);
+// Try to deserialize if response success
+match http_response {
+    Ok(response) => response.json::<SearchResult>(),
+    Err(e) => Err(e),
+}
+```
+
+We have to deal with two types of Error :
+ * error when doing HTTP call
+ * error when deserializing 
+
+When first error occurs we don't want to try deserialization. It breaks  the logic flow.
+        
+To avoid `spaghetti code` we can use `question mark  ?` operator :
+
+```rust
+let query = format!("http://hn.algolia.com/api/v1/search?query={}", keyword);
```

Comment:
[https://hn.algolia.com/api/v1/search?query={}](https://hn.algolia.com/api/v1/search?query=%7B%7D) - to avoid any problem with 301 redirect